File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/activity/HttpCallActivity.java
package com.prateekj.snooper.networksnooper.activity;

import android.Manifest;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayout.Tab;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.core.content.FileProvider;
import androidx.viewpager.widget.ViewPager;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.prateekj.snooper.R;
import com.prateekj.snooper.formatter.ResponseFormatterFactory;
import com.prateekj.snooper.infra.AppPermissionChecker;
import com.prateekj.snooper.infra.BackgroundTaskExecutor;
import com.prateekj.snooper.networksnooper.database.SnooperRepo;
import com.prateekj.snooper.networksnooper.fragment.HttpCallFragment;
import com.prateekj.snooper.networksnooper.fragment.HttpHeadersFragment;
import com.prateekj.snooper.networksnooper.helper.DataCopyHelper;
import com.prateekj.snooper.networksnooper.helper.HttpCallRenderer;
import com.prateekj.snooper.networksnooper.presenter.HttpCallPresenter;
import com.prateekj.snooper.networksnooper.views.HttpCallView;
import com.prateekj.snooper.utils.FileUtil;

import java.io.File;

import static android.content.Intent.ACTION_SEND;
import static android.content.Intent.EXTRA_STREAM;
import static android.content.Intent.EXTRA_SUBJECT;
import static android.content.Intent.FLAG_GRANT_READ_URI_PERMISSION;

public class HttpCallActivity extends SnooperBaseActivity implements HttpCallView {

  public static final String HTTP_CALL_ID = "HTTP_CALL_ID";
  public static final String HTTP_CALL_MODE = "HTTP_CALL_MODE";
  public static final int REQUEST_MODE = 1;
  public static final int RESPONSE_MODE = 2;
  public static final int ERROR_MODE = 3;
  public static final int WRITE_EXTERNAL_STORAGE_REQUEST_CODE = 1;
  private static final String LOGFILE_MIME_TYPE = "*/*";
  private HttpCallPresenter httpCallPresenter;
  private ViewPager pager;
  private HttpCallRenderer httpCallRenderer;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_http_call_detail);
    Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
    setSupportActionBar(toolbar);
    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    long httpCallId = getIntent().getLongExtra(HTTP_CALL_ID, 0);
    FileUtil fileUtil = new FileUtil();
    SnooperRepo repo = new SnooperRepo(this);
    BackgroundTaskExecutor backgroundTaskExecutor = new BackgroundTaskExecutor(this);
    DataCopyHelper dataCopyHelper = new DataCopyHelper(repo.findById(httpCallId), new ResponseFormatterFactory(), getResources());
    httpCallPresenter = new HttpCallPresenter(dataCopyHelper, repo.findById(httpCallId), this, fileUtil, backgroundTaskExecutor);
    boolean hasError = repo.findById(httpCallId).getError() != null;
    httpCallRenderer = new HttpCallRenderer(this, hasError);
    setupUi();
  }

  private void setupUi() {
    pager = (ViewPager) findViewById(R.id.pager);
    TabLayout tabLayout = (TabLayout) findViewById(R.id.tab_layout);
    for (HttpCallTab tab : httpCallRenderer.getTabs()) {
      tabLayout.addTab(tabLayout.newTab().setText(tab.getTabTitle()));
    }
    tabLayout.setTabGravity(TabLayout.GRAVITY_FILL);
    HttpCallPagerAdapter adapter = new HttpCallPagerAdapter(getSupportFragmentManager());
    pager.setAdapter(adapter);
    pager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));
    tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
      @Override
      public void onTabSelected(Tab tab) {
        pager.setCurrentItem(tab.getPosition());
      }

      @Override
      public void onTabUnselected(Tab tab) {

      }

      @Override
      public void onTabReselected(Tab tab) {

      }
    });
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.http_call_menu, menu);
    return true;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    if (item.getItemId() == R.id.copy_menu) {
      HttpCallTab currentTab = httpCallRenderer.getTabs().get(pager.getCurrentItem());
      httpCallPresenter.copyHttpCallBody(currentTab);
      return true;
    } else if (item.getItemId() == R.id.share_menu) {
      shareHttpCallData();
    }
    return super.onOptionsItemSelected(item);
  }

  @Override
  public void copyToClipboard(String data) {
    ClipboardManager clipboard = (ClipboardManager) this.getSystemService(CLIPBOARD_SERVICE);
    ClipData clip = ClipData.newPlainText("Copied", data);
    clipboard.setPrimaryClip(clip);
  }

  @Override
  public void shareData(String logFilePath) {
    File file = new File(logFilePath);
    Uri fileUri = FileProvider.getUriForFile(this, getApplicationContext().getPackageName() + ".provider", file);
    Intent intent = new Intent(ACTION_SEND);
    intent.setDataAndType(fileUri, LOGFILE_MIME_TYPE);
    intent.putExtra(EXTRA_SUBJECT, getString(R.string.mail_subject_share_logs));
    intent.putExtra(EXTRA_STREAM, fileUri);
    intent.addFlags(FLAG_GRANT_READ_URI_PERMISSION);
    Intent j = Intent.createChooser(intent, getString(R.string.chooser_title_share_logs));
    startActivity(j);
  }

  @Override
  public void showMessageShareNotAvailable() {
    Toast.makeText(this, R.string.permission_not_granted, Toast.LENGTH_SHORT).show();
  }

  private void shareHttpCallData() {
    appPermissionChecker.handlePermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, WRITE_EXTERNAL_STORAGE_REQUEST_CODE, new AppPermissionChecker.PermissionRequestCallBack() {
      @Override
      public void permissionGranted() {
        httpCallPresenter.shareHttpCallBody();
      }

      @Override
      public void permissionDenied() {
        httpCallPresenter.onPermissionDenied();
      }
    });
  }

  @Override
  public Fragment getResponseBodyFragment() {
    HttpCallFragment fragment = new HttpCallFragment();
    Bundle extras = getIntent().getExtras();
    extras.putInt(HTTP_CALL_MODE, RESPONSE_MODE);
    fragment.setArguments(extras);
    return fragment;
  }

  @Override
  public Fragment getRequestBodyFragment() {
    HttpCallFragment fragment = new HttpCallFragment();
    Bundle extras = getIntent().getExtras();
    extras.putInt(HTTP_CALL_MODE, REQUEST_MODE);
    fragment.setArguments(extras);
    return fragment;
  }

  @Override
  public Fragment getHeadersFragment() {
    HttpHeadersFragment fragment = new HttpHeadersFragment();
    fragment.setArguments(getIntent().getExtras());
    return fragment;
  }

  @Override
  public Fragment getExceptionFragment() {
    HttpCallFragment fragment = new HttpCallFragment();
    Bundle extras = getIntent().getExtras();
    extras.putInt(HTTP_CALL_MODE, ERROR_MODE);
    fragment.setArguments(extras);
    return fragment;
  }

  private class HttpCallPagerAdapter extends FragmentStatePagerAdapter {

    public HttpCallPagerAdapter(FragmentManager fm) {
      super(fm);
    }

    @Override
    public Fragment getItem(int position) {
      return httpCallRenderer.getFragment(position);
    }

    @Override
    public int getCount() {
      return httpCallRenderer.getTabs().size();
    }
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/activity/HttpCallActivity.kt
package com.prateekj.snooper.networksnooper.activity

import android.Manifest
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.Intent.ACTION_SEND
import android.content.Intent.EXTRA_STREAM
import android.content.Intent.EXTRA_SUBJECT
import android.content.Intent.FLAG_GRANT_READ_URI_PERMISSION
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.Tab
import com.prateekj.snooper.R
import com.prateekj.snooper.formatter.ResponseFormatterFactory
import com.prateekj.snooper.infra.AppPermissionChecker
import com.prateekj.snooper.infra.BackgroundTaskExecutor
import com.prateekj.snooper.networksnooper.database.SnooperRepo
import com.prateekj.snooper.networksnooper.fragment.HttpCallFragment
import com.prateekj.snooper.networksnooper.fragment.HttpHeadersFragment
import com.prateekj.snooper.networksnooper.helper.DataCopyHelper
import com.prateekj.snooper.networksnooper.helper.HttpCallRenderer
import com.prateekj.snooper.networksnooper.presenter.HttpCallPresenter
import com.prateekj.snooper.networksnooper.views.HttpCallView
import com.prateekj.snooper.utils.FileUtil
import kotlinx.android.synthetic.main.activity_http_call_detail.*
import java.io.File

class HttpCallActivity : SnooperBaseActivity(), HttpCallView {
  private var httpCallPresenter: HttpCallPresenter? = null
  private var httpCallRenderer: HttpCallRenderer? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_http_call_detail)
    setSupportActionBar(toolbar)
    supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    val httpCallId = intent.getLongExtra(HTTP_CALL_ID, 0)
    val fileUtil = FileUtil()
    val repo = SnooperRepo(this)
    val backgroundTaskExecutor = BackgroundTaskExecutor(this)
    val dataCopyHelper =
      DataCopyHelper(repo.findById(httpCallId), ResponseFormatterFactory(), resources)
    httpCallPresenter = HttpCallPresenter(
      dataCopyHelper,
      repo.findById(httpCallId),
      this,
      fileUtil,
      backgroundTaskExecutor
    )
    val hasError = repo.findById(httpCallId).error != null
    httpCallRenderer = HttpCallRenderer(this, hasError)
    setupUi()
  }

  private fun setupUi() {
    for (tab in httpCallRenderer!!.getTabs()) {
      tab_layout.addTab(tab_layout.newTab().setText(tab.tabTitle))
    }
    tab_layout.tabGravity = TabLayout.GRAVITY_FILL
    val adapter = HttpCallPagerAdapter(supportFragmentManager)
    pager!!.adapter = adapter
    pager!!.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tab_layout))
    tab_layout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
      override fun onTabSelected(tab: Tab) {
        pager!!.currentItem = tab.position
      }

      override fun onTabUnselected(tab: Tab) {

      }

      override fun onTabReselected(tab: Tab) {

      }
    })
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.http_call_menu, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    if (item.itemId == R.id.copy_menu) {
      val currentTab = httpCallRenderer!!.getTabs()[pager!!.currentItem]
      httpCallPresenter!!.copyHttpCallBody(currentTab)
      return true
    } else if (item.itemId == R.id.share_menu) {
      shareHttpCallData()
    }
    return super.onOptionsItemSelected(item)
  }

  override fun copyToClipboard(data: String) {
    val clipboard = this.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    val clip = ClipData.newPlainText("Copied", data)
    clipboard.primaryClip = clip
  }

  override fun shareData(logFilePath: String) {
    val file = File(logFilePath)
    val fileUri =
      FileProvider.getUriForFile(this, applicationContext.packageName + ".provider", file)
    val intent = Intent(ACTION_SEND)
    intent.setDataAndType(fileUri, LOGFILE_MIME_TYPE)
    intent.putExtra(EXTRA_SUBJECT, getString(R.string.mail_subject_share_logs))
    intent.putExtra(EXTRA_STREAM, fileUri)
    intent.addFlags(FLAG_GRANT_READ_URI_PERMISSION)
    val j = Intent.createChooser(intent, getString(R.string.chooser_title_share_logs))
    startActivity(j)
  }

  override fun showMessageShareNotAvailable() {
    Toast.makeText(this, R.string.permission_not_granted, Toast.LENGTH_SHORT).show()
  }

  private fun shareHttpCallData() {
    appPermissionChecker.handlePermission(
      Manifest.permission.WRITE_EXTERNAL_STORAGE,
      WRITE_EXTERNAL_STORAGE_REQUEST_CODE,
      object : AppPermissionChecker.PermissionRequestCallBack {
        override fun permissionGranted() {
          httpCallPresenter!!.shareHttpCallBody()
        }

        override fun permissionDenied() {
          httpCallPresenter!!.onPermissionDenied()
        }
      })
  }

  override fun getResponseBodyFragment(): Fragment {
    val fragment = HttpCallFragment()
    val extras = intent.extras
    extras!!.putInt(HTTP_CALL_MODE, RESPONSE_MODE)
    fragment.arguments = extras
    return fragment
  }

  override fun getRequestBodyFragment(): Fragment {
    val fragment = HttpCallFragment()
    val extras = intent.extras
    extras!!.putInt(HTTP_CALL_MODE, REQUEST_MODE)
    fragment.arguments = extras
    return fragment
  }

  override fun getHeadersFragment(): Fragment {
    val fragment = HttpHeadersFragment()
    fragment.arguments = intent.extras
    return fragment
  }

  override fun getExceptionFragment(): Fragment {
    val fragment = HttpCallFragment()
    val extras = intent.extras
    extras!!.putInt(HTTP_CALL_MODE, ERROR_MODE)
    fragment.arguments = extras
    return fragment
  }

  private inner class HttpCallPagerAdapter(fm: FragmentManager) : FragmentStatePagerAdapter(fm) {
    override fun getItem(position: Int): Fragment {
      return httpCallRenderer!!.getFragment(position)
    }

    override fun getCount(): Int {
      return httpCallRenderer!!.getTabs().size
    }
  }

  companion object {
    const val HTTP_CALL_ID = "HTTP_CALL_ID"
    const val HTTP_CALL_MODE = "HTTP_CALL_MODE"
    const val REQUEST_MODE = 1
    const val RESPONSE_MODE = 2
    const val ERROR_MODE = 3
    const val WRITE_EXTERNAL_STORAGE_REQUEST_CODE = 1
    private const val LOGFILE_MIME_TYPE = "*/*"
  }
}
